src/main/java/org/bitsofinfo/hazelcast/discovery/docker/swarm/SwarmDiscoveryUtil.java

```diff
@@ -322,49 +325,48 @@ public void setHazelcastPeerPort(Integer hazelcastPeerPort) {
 			Set<DiscoveredContainer> discoveredContainers = new HashSet<DiscoveredContainer>();
 
 			// the relevant networks we are looking for
-			Map<String,Network> relevantNetIds2Networks = new TreeMap<String,Network>(String.CASE_INSENSITIVE_ORDER);
+			Map<String, Network> relevantNetIds2Networks = new TreeMap<String, Network>(String.CASE_INSENSITIVE_ORDER);
 
 			// build list of relevant networkIds -> Network objects we care about
 			for (String dockerNetworkName : this.getDockerNetworkNames()) {
 				List<Network> networks = docker.listNetworks(ListNetworksParam.byNetworkName(dockerNetworkName));
 				for (Network network : networks) {
-					relevantNetIds2Networks.put(network.id(),network);
-					logger.info("SwarmDiscoveryUtil["+this.context+"] Found relevant docker network: " + network.name() +"["+ network.id() + "]");
+					relevantNetIds2Networks.put(network.id(), network);
+					logger.info("SwarmDiscoveryUtil[" + this.context + "] Found relevant docker network: " + network.name() + "[" + network.id() + "]");
 				}
 			}
-			
-	        if(relevantNetIds2Networks.size() == 0) {
-	            logger.warning("SwarmDiscoveryUtil["+this.context+"] Did not find relevant docker network for: " + this.getDockerNetworkNames());
-	        }
+
+			if (relevantNetIds2Networks.size() == 0) {
+				logger.warning("SwarmDiscoveryUtil[" + this.context + "] Did not find relevant docker network for: " + this.getDockerNetworkNames());
+			}
 
 			// Collect all relevant containers for services with services-names on the relevant networks
 			for (String dockerServiceName : this.getDockerServiceNames()) {
-				logger.info("SwarmDiscoveryUtil["+this.context+"] Invoking criteria-based container discovery for dockerServiceName=" + dockerServiceName);
+				logger.info("SwarmDiscoveryUtil[" + this.context + "] Invoking criteria-based container discovery for dockerServiceName=" + dockerServiceName);
 				discoveredContainers.addAll(
 						discoverContainersViaCriteria(docker,
-										   relevantNetIds2Networks,
-										   Criteria.builder().serviceName(dockerServiceName).build()));
+								relevantNetIds2Networks,
+								Criteria.builder().serviceName(dockerServiceName).build(),
+								new NameBasedServiceFilter(dockerServiceName)));
```

Comment:
This has to be wrapped via a new configuration option. Only if it is set, then apply this new NameBasedServiceFilter forced match, otherwise do not (i.e. the behavior as it works today)

I get the intent but there are others using this library and they may be intentionally or unintentionally benefiting from the prefix based matching behavior that the underlying spotify and docker service apis provide. This change for your specific situation must not impact backwards compatibility.

Reply:
Configured as: strict-docker-service-name-comparison / strictDockerServiceNameComparison